src/protocols/mcsp/MessageCounterManager.cpp

```diff
@@ -194,97 +255,90 @@ CHIP_ERROR MessageCounterSyncMgr::AddToReceiveTable(const PacketHeader & packetH
+
     // Find all receive entries matching peerNodeId.  Note that everything in
     // this table was using an application group key; that's why it was added.
     for (ReceiveTableEntry & entry : mReceiveTable)
     {
-        if (!entry.msgBuf.IsNull() && entry.session.GetPeerNodeId() == peerNodeId)
+        if (entry.peerNodeId == peerNodeId)
         {
             // Reprocess message.
-            mExchangeMgr->HandleGroupMessageReceived(entry.packetHeader, entry.payloadHeader, entry.session,
-                                                     std::move(entry.msgBuf));
+            secureSessionMgr->OnMessageReceived(entry.packetHeader, entry.peerAddress, std::move(entry.msgBuf));
 
             // Explicitly free any buffer owned by this handle.  The
             // HandleGroupMessageReceived() call should really handle this, but
             // just in case it messes up we don't want to get confused about
             // wheter the entry is in use.
-            entry.msgBuf = nullptr;
+            entry.msgBuf     = nullptr;
+            entry.peerNodeId = kUndefinedNodeId;
         }
     }
 }
 
 // Create and initialize new exchange for the message counter synchronization request/response messages.
-CHIP_ERROR MessageCounterSyncMgr::NewMsgCounterSyncExchange(SecureSessionHandle session,
+CHIP_ERROR MessageCounterManager::NewMsgCounterSyncExchange(SecureSessionHandle session,
                                                             Messaging::ExchangeContext *& exchangeContext)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
-    // Message counter synchronization protocol is only applicable for application group keys.
-    VerifyOrReturnError(ChipKeyId::IsAppGroupKey(session.GetPeerKeyId()), err = CHIP_ERROR_INVALID_ARGUMENT);
-
     // Create new exchange context.
     exchangeContext = mExchangeMgr->NewContext(session, this);
     VerifyOrReturnError(exchangeContext != nullptr, err = CHIP_ERROR_NO_MEMORY);
 
     return err;
 }
 
-CHIP_ERROR MessageCounterSyncMgr::SendMsgCounterSyncReq(SecureSessionHandle session)
+CHIP_ERROR MessageCounterManager::SendMsgCounterSyncReq(SecureSessionHandle session, Transport::PeerConnectionState * state)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
     Messaging::ExchangeContext * exchangeContext = nullptr;
-    Transport::PeerConnectionState * state       = nullptr;
     System::PacketBufferHandle msgBuf;
     Messaging::SendFlags sendFlags;
-    uint8_t challenge[kMsgCounterChallengeSize];
-
-    state = mExchangeMgr->GetSessionMgr()->GetPeerConnectionState(session);
-    VerifyOrExit(state != nullptr, err = CHIP_ERROR_NOT_CONNECTED);
+    std::array<uint8_t, kChallengeSize> challenge;
 
     // Create and initialize new exchange.
     err = NewMsgCounterSyncExchange(session, exchangeContext);
     SuccessOrExit(err);
 
     // Allocate a buffer for the null message.
-    msgBuf = MessagePacketBuffer::New(kMsgCounterChallengeSize);
+    msgBuf = MessagePacketBuffer::New(kChallengeSize + 16);
```

Comment:
please add a constant for 16 

needs a name to understand why the extra space

Reply:
@andy31415 Here `+16` is the extra space at the end of the packet for MAC. I have tried without `+16` here, but got out of memory when encoding the packet, because there are no space to put the MAC. I'm not sure if I'm using the `PacketBuffer` correctly, but I expect that the protocol or application doesn't need to reserve the space for MAC.